### PR TITLE
expression: rewrite builtin function: PERIOD_ADD, PERIOD_DIFF

### PR DESCRIPTION
**Pull request by zz-jason**
to #4080

### CODE REVIEW

**Comment by zz-jason** [comment]
@breeswish @XuHuaiyu PTAL

**Comment by breeswish** [comment]
LGTM

File: expression/builtin_time.go [hunk 1]
```diff
@@ -2546,52 +2546,71 @@ func (c *periodAddFunctionClass) getFunction(args []Expression, ctx context.Cont
...
+		return nil, errors.Trace(err)
+	}
+	bf.tp.Flen = 6
+	sig := &builtinPeriodAddSig{baseIntBuiltinFunc{bf}}
 	return sig.setSelf(sig), nil
 }
 
-type builtinPeriodAddSig struct {
-	baseBuiltinFunc
+func period2Month(period uint64) uint64 {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: 1. add a comment for `period`
2. add a comment for this function.

File: expression/builtin_time.go [hunk 2]
```diff
@@ -2546,52 +2546,71 @@ func (c *periodAddFunctionClass) getFunction(args []Expression, ctx context.Cont
...
+	return year*12 + month - 1
 }
 
-// eval evals a builtinPeriodAddSig.
-// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_period-add
-func (b *builtinPeriodAddSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func month2Period(month uint64) uint64 {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: add a comment for this function.

File: expression/integration_test.go [hunk 3]
```diff
@@ -1148,6 +1148,18 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
...
 	result.Check(testkit.Rows("123456 10 <nil> <nil>"))
 
+	// for period_add
+	result = tk.MustQuery(`SELECT period_add(191, 2), period_add(191, -2), period_add(0, 20), period_add(0, 0);`)
+	result.Check(testkit.Rows("200809 200805 0 0"))
+	result = tk.MustQuery(`SELECT period_add(NULL, 2), period_add(-191, NULL), period_add(NULL, NULL), period_add(12.09, -2), period_add("21aa", "11aa"), period_add("", "");`)
+	result.Check(testkit.Rows("<nil> <nil> <nil> 200010 200208 0"))
+
+	// for period_diff
+	result = tk.MustQuery(`SELECT period_diff(191, 2), period_diff(0, 0), period_diff(191, 191);`)
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: add test case for period_diff(191, -2);

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Pull request merged**